x-pack/legacy/plugins/reporting/server/legacy.ts

```diff
@@ -4,35 +4,87 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 import { Legacy } from 'kibana';
-import { PluginInitializerContext } from 'src/core/server';
+import { get } from 'lodash';
+import { take } from 'rxjs/operators';
+import { CoreSetup, PluginInitializerContext } from 'src/core/server';
+import { ConfigType, PluginsSetup } from '../../../../plugins/reporting/server';
 import { SecurityPluginSetup } from '../../../../plugins/security/server';
 import { ReportingPluginSpecOptions } from '../types';
 import { plugin } from './index';
-import { LegacySetup, ReportingStartDeps } from './types';
+import { LegacySetup, ReportingConfig, ReportingStartDeps } from './types';
 
 const buildLegacyDependencies = (
+  coreSetup: CoreSetup,
   server: Legacy.Server,
   reportingPlugin: ReportingPluginSpecOptions
-): LegacySetup => ({
-  config: server.config,
-  info: server.info,
-  route: server.route.bind(server),
-  plugins: {
-    elasticsearch: server.plugins.elasticsearch,
-    xpack_main: server.plugins.xpack_main,
-    reporting: reportingPlugin,
-  },
-});
+): LegacySetup => {
+  return {
+    route: server.route.bind(server),
+    plugins: {
+      xpack_main: server.plugins.xpack_main,
+      reporting: reportingPlugin,
+    },
+  };
+};
+
+interface KbnServerConfig {
+  path: { data: string };
+  server: {
+    basePath: string;
+    host: string;
+    name: string;
+    port: number;
+    protocol: string;
+    uuid: string;
+  };
+}
+
+const buildLegacyConfig = (
+  coreSetup: CoreSetup,
+  server: Legacy.Server,
+  reportingConfig: ConfigType
+): ReportingConfig => {
+  const config = server.config();
+  const { http } = coreSetup;
+  const serverInfo = http.getServerInfo();
+
+  const kbnConfig: KbnServerConfig = {
+    path: {
+      data: config.get('path.data'), // FIXME get from new platform. See https://github.com/elastic/kibana/issues/56760#issuecomment-582008704
+    },
+    server: {
+      basePath: coreSetup.http.basePath.serverBasePath,
+      host: serverInfo.host,
+      name: serverInfo.name,
+      port: serverInfo.port,
+      uuid: coreSetup.uuid.getInstanceUuid(),
+      protocol: serverInfo.protocol,
+    },
+  };
+
+  return {
+    get: (key: string) => get<ReportingConfig>(reportingConfig, key),
```

Comment:
Note: the type information of the data at `key` is getting lost from this function. I want to figure out how to return the type as well so any calling code stays strict.

Reply:
we can do this with the array syntax of `get`, and use differnt overloaded types of `get` based on the number of key params:
```
interface Config<BaseType> {
  get<Key1 extends keyof BaseType>(key1: Key1): BaseType[Key1];
  get<Key1 extends keyof BaseType, Key2 extends keyof BaseType[Key1]>(key1: Key1, key2: Key2): BaseType[Key1][Key2];
  // etc
}
```